File: 389-find-the-difference.rb
def find_the_difference(s, t)
    hash = {};
    s.chars.each do |char|
        if hash[char] == nil
            hash[char] = 1
        else
            hash[char] += 1
        end
    end

    t.chars.each do |char|
        if (hash[char] == nil || hash[char] <= 0)
            return char
        end
        hash[char] -= 1
    end

end

def find_the_difference(s, t)
    s = s << t
    result = 0;
    s.chars.each do |char|
        result ^= char.ord
    end
    result.chr
end

File: 389-find-the-difference.js
var findTheDifference = function(s, t) {
    let hash = {}
    for (let i = 0; i < s.length; i++) {
        hash[s[i]] = (hash[s[i]] + 1) || 1;
    }
    for (let i = 0; i < t.length; i++) {
        if (hash[t[i]] === undefined || hash[t[i]] <= 0) {
            return t[i];
        }
        hash[t[i]]--;
    }
};

var findTheDifference = function(s, t) {
    let combinedStr = s + t;
    let result = 0;
    for (let i = 0; i < combinedStr.length; i++) {
        result ^= combinedStr.charCodeAt(i);
    }
    return String.fromCharCode(result);
};
